### PR TITLE
Add MaxZonesWithUnavailablePods for the update process. Fixes #1444.

### LINKED ISSUE #1444: Implement Max Unavailable
### What would you like to be added/changed?

We should implement a `Max Unavailable` setting similar to what a Kubernetes Deployment provides: https://kubernetes.io/docs/concepts/workloads/controllers/deployment/#max-unavailable to prevent cases where the operator is deleting Pods and the new Pods are stuck in Pending. See https://github.com/FoundationDB/fdb-kubernetes-operator/pull/1236#discussion_r1032580981 for some more context.

### PR DESCRIPTION
**Pull request by manfontan**
# Description

`MaxZonesWithUnavailablePods` option limits the number of zones (fault domains) with unavailable pods during an update. By default it is disabled

fixes #1444 

duplicates #1525 

## Type of change

- New feature (non-breaking change which adds functionality)

## Discussion

*Are there any design details that you would like to discuss further?
NO

## Testing

*Please describe the tests that you ran to verify your changes. Unit tests?
Manual testing?
Unit tests
Manual testing 

Replace FDB base image for all log,storage and stateless:

Fault Tolerance was always above 1 and replicas were always >=2.

Screenshot 2023-06-02 at 17 25 07

The feature prevented the operator from deleting pods too fast 36 times with messages like:

```
fdb-kubernetes-operator-controller-manager-66dbb578c-hzrmt manager {"level":"info","ts":1685722511.3212197,"logger":"controller","msg":"Skip process group for update, the number zones with unavailable Pods equals the limit but the process group does not belong to a zone with unavailable Pods","namespace":"timeseries","cluster":"foundationdb-cluster","reconciler":"updatePods","processGroupID":"storage-8","maxZonesWithUnavailablePods":1,"faultDomain":"fdb1-26081535-vmss0000z2","faultDomainsWithUnavailablePods":{"fdb1-26081535-vmss0000yv":{}}}
```

## Documentation

*Did you update relevant documentation within this repository? Yes

*If this change is adding new functionality, do we need to describe it in our user manual? Yes

*If this change is adding or removing subreconcilers, have we updated the core technical design doc to reflect that?N/A

*If this change is adding new safety checks or new potential failure modes, have we documented and how to debug potential issues?*

## Follow-up

*Are there any follow-up issues that we should pursue in the future? No

*Does this introduce new defaults that we should re-evaluate in the future? Yes. By default the max number of unavailable pods during updates is not limited. In the future it could be set to a specific number.

### CODE REVIEW

File: controllers/update_pods.go
Comment (johscheuer) [inline controllers/update_pods.go:96]: Any reason why a process group marked for removal is counted as unavailable? I would have assumed that unavailable means the Pod is not existing or in a not running state?

File: controllers/update_pods.go
Comment (johscheuer) [inline controllers/update_pods.go:110]: I think increasing the number of `unavailablePods` is wrong here.

File: api/v1beta2/foundationdbcluster_types.go
Comment (johscheuer) [inline api/v1beta2/foundationdbcluster_types.go:225]: There is an `intstr` type that we might want to use, to have the possibility to use a percentage instead of using an integer,

**Review by johscheuer — CHANGES_REQUESTED** [review]
Could you please add tests that show that the newly added setting is actually behaving as expected and preventing the operator from taking down more than thee specified Pods?

File: config/rbac/role.yaml
Comment (manfontan) [inline config/rbac/role.yaml:3]: not sure why this changed. I did not do it 🤔

File: controllers/update_pods.go
Comment (manfontan) [inline controllers/update_pods.go:99]: I have created a dedicated validation since SkipProcessGroups uses the ignore pending time that defaults to 5min. 
And I think it is simpler to check it separately than having to set two values to get the functionality. Specially because the default ignore pending time value results in Pending Pods not being counted unless they have been pending for more than 5 min.

File: config/samples/deployment.yaml
Comment (manfontan) [inline config/samples/deployment.yaml:153]: ignore this change I will revert it.

File: controllers/update_pods.go
Comment (johscheuer) [inline controllers/update_pods.go:203]: Should we print here the number of unavailable Pods?

File: controllers/update_pods_test.go
Comment (johscheuer) [inline controllers/update_pods_test.go:264]: This case is never hit right? The initial value of `numPendingPods` will be 0 and before we do the check here we are increasing the value by 1, so it will always be > 0 and all process groups will get this condition? Was the idea of `numPendingPods` to be used for debugging?

File: controllers/update_pods_test.go
Comment (johscheuer) [inline controllers/update_pods_test.go:278]:
```suggestion
		When("max unavailable pods is set with an int value and there are two process groups with pods in pending status", func() {
```

File: controllers/update_pods_test.go
Comment (johscheuer) [inline controllers/update_pods_test.go:253]: Can we add a test case where we actually update some Pods even if some of them are unavailable?

File: controllers/update_pods.go
Comment (johscheuer) [inline controllers/update_pods.go:205]: I believe the current logic is not doing what it proposes: `MaxUnavailablePods defines the maximum number or percent of pods that can be unavailable during the update process.`

But we are actually only checking if `unavailablePods >= maxUnavailablePods` before doing the updates. So we could hit a case where `maxUnavailablePods` is `5` and `unavailablePods` is `4`, this would mean `unavailablePods >= maxUnavailablePods` is false and we have an update budget of 1 Pod, but in the current implementation we are not verifying if the `updates` will only update `1` Pod and not more (otherwise we have more unavailable Pods than defined in `maxUnavailablePods`)

**Review by johscheuer — CHANGES_REQUESTED** [review]
Most of the logic looks fine, we just have to make sure we are not updating more Pods then we have a "budget" for.

File: controllers/update_pods_test.go
Comment (manfontan) [reply controllers/update_pods_test.go:264]: The idea here was to set the PodPending condition for a certain number (8 in this case) of process groups. Once the desired number is reached `numPendingPods` we skip setting the PodsPending condition for the remaining process groups.

Does that make sense?

File: controllers/update_pods.go
Comment (johscheuer) [inline controllers/update_pods.go:207]: I believe this calculation is still wrong. Assume a user sets `maxUnavailablePods` to `1` to make sure only 1 Pod is unavailable at a time (at least from planned activities) and now the user makes a change to the cluster that requires the operator to update all Pods in the cluster. For simplicity assume the cluster has 3 Pods, in this case the operator will never be able to make progress as `(unavailablePods+numPodsToUpdate) >= maxUnavailablePods` will be `3 >= 1` which will be blocking forever.

I believe a better approach is to return the updates map and the number of Pods that can be updated, in the case above that would be 1 if all Pods are available. If the updates map has at lease one entry but the Pods that can be updated is 0 (or a negative number) the operator should do a (delayed) requeue. If the number of Pods that can be updated is greater than 0 the operator should prefer to update zones that have less or equal the number of Pods that can be updated (to make sure we update the zone at once) if not the operator should update as many Pods as possible from the zone.

That way we make sure the operator is able to make progress in the different cases.

File: controllers/update_pods_test.go
Comment (johscheuer) [inline controllers/update_pods_test.go:307]: Can we use a more reasonable value here to e.g. allow only to update 1 Pod. Basically the value 100 in this test case is the same as setting the limit to 0 (not having any limit) as the Pods created by this cluster is much smaller than 100.

File: controllers/update_pods.go
Comment (johscheuer) [inline controllers/update_pods.go:93]: I believe we should count Pods that have a deletionTimestamp !=0 as unavailable, what do you think?

File: controllers/update_pods_test.go
Comment (johscheuer) [reply controllers/update_pods_test.go:264]: That makes sense, I just was confused what the benefit of the previous check was.

**Review by johscheuer — CHANGES_REQUESTED** [review]

File: controllers/update_pods.go
Comment (manfontan) [reply controllers/update_pods.go]: That makes total sense. I like the idea. I hope I got it right this time :D

**Review by johscheuer — COMMENTED** [review]
Once we merge the new `TrimUpdatedToMaxPodsToUpdate` with the `getPodsToDelete` this PR is good. Do you mind to add a section about the new setting in https://github.com/FoundationDB/fdb-kubernetes-operator/blob/main/docs/manual/replacements_and_deletions.md? Could you also create an issue to build a dedicated e2e test for this new feature?

**Comment by manfontan** [comment]
> Once we merge the new `TrimUpdatedToMaxPodsToUpdate` with the `getPodsToDelete` this PR is good. Do you mind to add a section about the new setting in https://github.com/FoundationDB/fdb-kubernetes-operator/blob/main/docs/manual/replacements_and_deletions.md? Could you also create an issue to build a dedicated e2e test for this new feature?

I will add a section in the doc. 

About merging `TrimUpdatedToMaxPodsToUpdate` I don't think it is necessary since the new `getPodsToDelete` logic already does the same thing. 

Depending on the deletion Mode:
- PodUpdateModeAll: maxPodsToUpdate is ignored.
- PodUpdateModeProcessGroup: we only delete pods if maxPodsToUpdate>=1
- PodUpdateModeZone:
  - If the zone process count < maxPodsToUpdate we pick the entire zone to delete.
  - If the zone is > maxPodsToUpdate we trim the slice to get only the first maxPodsToUpdate processes
- PodUpdateModeNone: deletes nothing

Does that make sense?

File: api/v1beta2/foundationdbcluster_types.go
Comment (johscheuer) [inline api/v1beta2/foundationdbcluster_types.go:226]: I think in this case the percentage doesn't make sense as the zones can be quite dynamic, so converting this to an `*int` would be better.

File: controllers/update_pods.go
Comment (johscheuer) [inline controllers/update_pods.go:93]: This calculation is not correct. As I mentioned it makes sense to use an integer here instead of trying to calculate the percentage.

File: controllers/update_pods.go
Comment (johscheuer) [inline controllers/update_pods.go:86]:
```suggestion
	faultDomainsWithUnavailablePods := make(map[string]fdbv1beta2{})
```

We don't need to count the Pods, as we don't care about the number of unavailable Pods but the number of fault domains containing at least one unavailable Pod.

**Review by johscheuer — COMMENTED** [review]
A few comments around some thing that can be simplified otherwise the changes look good. Have you done any tests besides the unit tests?

**Comment by manfontan** [comment]
I will deploy the changes in dev tomorrow and perform some testing with the latest updates.

File: api/v1beta2/foundationdbcluster_types.go
Comment (johscheuer) [inline api/v1beta2/foundationdbcluster_types.go:2537]:
```suggestion
// GetMaxZonesWithUnavailablePods returns the maximum number of zones that can have unavailable pods.
```

File: api/v1beta2/foundationdbcluster_types.go
Comment (johscheuer) [inline api/v1beta2/foundationdbcluster_types.go:2539]:
```suggestion
	return pointer.IntDeref(cluster.Spec.MaxZonesWithUnavailablePods, math.MaxInt)
```

If we just set this value to the maximum integer value we don't have to make any special cases around `0`.

File: controllers/update_pods.go
Comment (johscheuer) [inline controllers/update_pods.go:159]: When we change the default to `math.MaxInt` we can remove this check.

File: api/v1beta2/foundationdbcluster_types.go
Comment (johscheuer) [inline api/v1beta2/foundationdbcluster_types.go:224]:
```suggestion
	// When unset, there is no limit on the number of zones with unavailable pods.
```

**Review by johscheuer — COMMENTED** [review]
Once the lint issue is fixed we are good 👍 One last proposal on using `math.MaxInt` instead of `0` to represent no limit on zones with unavailable Pods, this will remove the need for any additional checks for `0`.

**Review by johscheuer — APPROVED** [review]
LGTM 👍 Once you've verified those changes in dev we can merge them. Thanks!

**Review by johscheuer — APPROVED** [review]

**Pull request merged**
